src/toil/lib/threading.py

```diff
@@ -335,9 +333,222 @@ def process_name_exists(workDir, name):
             except:
                 pass
     
+# Similar to the process naming system above, we define a global mutex system
+# for critical sections, based just around file locks.
+@contextmanager
+def global_mutex(workDir, mutex):
+    """
+    Context manager that locks a mutex. The mutex is identified by the given
+    name, and scoped to the given directory. Works across all containers that
+    have access to the given diectory. Mutexes held by dead processes are
+    automatically released.
+    
+    Only works between processes, NOT between threads.
+    
+    :param str workDir: The Toil work directory. Defines the shared namespace.
+    :param str mutex: Mutex to lock. Must be a permissible path component.
+    """
+    
+    # Define a filename
+    lock_filename = os.path.join(workDir, 'toil-mutex-' + mutex)
+    
+    log.debug('PID %d acquiring mutex %s', os.getpid(), lock_filename)
+    
+    # We can't just create/open and lock a file, because when we clean up
+    # there's a race where someone can open the file before we unlink it and
+    # get a lock on the deleted file.
+    
+    while True:
+        # Try to create the file, ignoring if it exists or not.
+        fd = os.open(lock_filename, os.O_CREAT | os.O_WRONLY)
+        
+        # Wait until we can exclusively lock it.
+        fcntl.lockf(fd, fcntl.LOCK_EX)
+        
+        # Holding the lock, make sure we are looking at the same file on disk still.
+        fd_stats = os.fstat(fd)
+        try:
+            path_stats = os.stat(lock_filename)
+        except FileNotFoundError:
+            path_stats = None
+        
+        if path_stats is None or fd_stats.st_dev != path_stats.st_dev or fd_stats.st_ino != path_stats.st_ino:
+            # The file we have a lock on is not the file linked to the name (if
+            # any). This usually happens, because before someone releases a
+            # lock, they delete the file. Go back and contend again. TODO: This
+            # allows a lot of queue jumping on our mutex.
+            fcntl.lockf(fd, fcntl.LOCK_UN)
+            os.close(fd)
+            continue
```

Comment:
We could say something like "the default timeout for waiting for a mutex is X minutes" or something. But I figured that the probability of someone actually managing to starve a process for X minutes and coming back to complain in a bug report was pretty high, especially if we ever re-use this for something else. Plus, generally you expect a mutex to block until it successfully locks; that's its job.

There is some probability that a worker manages to hang in some deferred function cleanup forever and blocks all the other workers. But no particular timeout is going to be "correct" either.

Reply:
Hmmm... alright, I'll defer to you on this one since it is trying to be OS-like and therefore could reasonably be expected to block infinitely.  In general I like to have an escape hatch, but I can't really think of a good default timeout either.  Something to keep in mind though if we ever see a cached kube job somehow frozen on one worker.